analysis/src/machine_check.rs

```diff
@@ -13,31 +16,80 @@ use powdr_ast::{
         self,
         asm::{
             self, ASMModule, ASMProgram, AbsoluteSymbolPath, AssignmentRegister, FunctionStatement,
-            Instruction, LinkDeclaration, MachineProperties, MachineStatement, ModuleStatement,
-            RegisterFlag, SymbolDefinition,
+            Instruction, LinkDeclaration, MachineProperties, MachineStatement, Module,
+            ModuleStatement, RegisterFlag, SymbolDefinition, SymbolValue,
         },
+        EnumDeclaration, TraitDeclaration,
     },
 };
 
 /// Verifies certain properties of each machine and constructs the Machine objects.
 /// Also transfers generic PIL definitions but does not verify anything about them.
 pub fn check(file: ASMProgram) -> Result<AnalysisASMFile, Vec<String>> {
     let ctx = AbsoluteSymbolPath::default();
-    let machines = TypeChecker::default().check_module(file.main, &ctx)?;
+    let checker = TypeChecker::new(file.main);
+    checker.check_path(&ctx)?;
     Ok(AnalysisASMFile {
-        items: machines.into_iter().collect(),
+        items: checker.analyzed.into_inner(),
     })
 }
 
-#[derive(Default)]
-struct TypeChecker {}
+struct TypeChecker {
+    /// The parsed AST as a single [SymbolValue::Module]
+    parsed: SymbolValue,
+    /// The already checked paths
+    checked: RefCell<HashSet<AbsoluteSymbolPath>>,
+    /// The checked items
+    /// We use [RefCell] because we want to mutate this while borrowing from `parsed`
+    analyzed: RefCell<BTreeMap<AbsoluteSymbolPath, Item>>,
+}
 
 impl TypeChecker {
+    fn new(root: ASMModule) -> Self {
+        Self {
+            parsed: SymbolValue::Module(Module::Local(root)),
+            checked: Default::default(),
+            analyzed: Default::default(),
+        }
+    }
+
+    fn check_path(&self, ctx: &AbsoluteSymbolPath) -> Result<(), Vec<String>> {
+        // check if the path is already checked
+        if self.checked.borrow().contains(ctx) {
+            return Ok(());
+        }
+
+        let symbol = ctx.parts().fold(&self.parsed, |acc, part| match acc {
+            SymbolValue::Module(Module::Local(m)) => m
+                .statements
+                .iter()
+                .find_map(|s| match s {
+                    ModuleStatement::SymbolDefinition(s) if s.name == part => Some(&s.value),
+                    _ => None,
+                })
+                .unwrap(),
+            s => s,
+        });
+
+        let res = match symbol {
+            SymbolValue::Module(Module::Local(module)) => self.check_module(module, ctx),
+            SymbolValue::Machine(machine) => self.check_machine_type(machine.clone(), ctx),
+            SymbolValue::Expression(e) => self.check_expression(e.clone(), ctx),
+            SymbolValue::TypeDeclaration(d) => self.check_type_declaration(d.clone(), ctx),
+            SymbolValue::TraitDeclaration(d) => self.check_trait_declaration(d.clone(), ctx),
+            _ => unreachable!(),
+        };
+
+        self.checked.borrow_mut().insert(ctx.clone());
```

Comment:
If an item references itself, and we only insert it into checked here, I think we might get infinite recursion.

Reply:
Good point but this is already the case in airgen currently (not infinite recursion but infinite loop) for a machine which has an instance of itself as a submachine. Can you think of other examples which this PR would break? If not, I think we should have a separate PR.